sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -293,6 +299,34 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     }
   }
 
+  /**
+   * Move the cast to the literal side, because we can only get the minimum value of timestamp,
+   * so some BinaryComparison needs to be changed,
+   * such as CAST(ts AS date) > DATE '2023-01-01' ===> ts >= TIMESTAMP '2023-01-02 00:00:00'
```

Comment:
If the date part of `ts` is greater than a date, then `ts` must be greater than the timestamp form of this date, right?

Reply:
Yes. It is.

Reply:
then why do we change `>` to `>=` after the conversion?

Reply:
Suppose we have:
```
CAST('2023-01-02 00:00:00' AS DATE) > DATE '2023-01-01'
```

After the rule, it'll become:
```
'2023-01-02 00:00:00' >= '2023-01-02 00:00:00'
```

if changing to `>` this no longer evaluate to true.

I think this is similar to line 281 where we handle the case where a numeric value is rounded up after converting to the `fromType`.

Reply:
Yes, it should be `'2023-01-02 00:00:00' > '2023-01-01 23:59:59.999999'` if do not change `>` to `>=`.

Reply:
It is because it is easy to get the start of the day timestamp, not easy to get the end of the day timestamp.
```sql
spark-sql> select cast(date'2023-01-01' as timestamp);
2023-01-01 00:00:00
```
vs
```sql
spark-sql> select cast(concat(cast(date'2023-01-01' as string), ' 23:59:59.999999') as timestamp);
2023-01-01 23:59:59.999999
```


It will truncate if casting from long type to timestamp type:
```sql
spark-sql> select cast(cast(cast(date'2023-01-01' as timestamp) as long) + (24* 60 * 60 - 1) as timestamp);
2023-01-01 23:59:59
```